### PR TITLE
Fixed #30913 -- Added support for covering indexes on PostgreSQL 11+.

### PR DESCRIPTION
**Pull request by hannseman**
Ticket: https://code.djangoproject.com/ticket/30913

### CODE REVIEW

File: django/db/backends/base/schema.py [hunk 1]
```diff
@@ -942,16 +942,27 @@ def _get_index_tablespace_sql(self, model, fields, db_tablespace=None):
...
 
+    def _get_include_sql(self, model, include_fields):
+        if not include_fields:
+            return ''
+        include_columns = Columns(
+            model._meta.db_table,
+            [field.column for field in include_fields],
+            self.quote_name
+        )
+        return ' INCLUDE (%s)' % include_columns if include_columns else ''
```
Comment (charettes) [inline django/db/backends/base/schema.py]: You'll want to return a `Statement` instead of interpolating here so aliases can be repointed if an `AddIndex` operation is included in a migration that also includes a rename or deletion operation and the index creation needs to be deferred.

```python
return Statement(
    ' INCLUDE (%(columns)s)',
    columns=columns
) if columns else ''
```

File: django/db/backends/postgresql/features.py [hunk 2]
```diff
@@ -64,10 +64,16 @@ def is_postgresql_9_6(self):
...
+
     @cached_property
     def is_postgresql_12(self):
         return self.connection.pg_version >= 120000
 
     has_brin_autosummarize = property(operator.attrgetter('is_postgresql_10'))
     has_phraseto_tsquery = property(operator.attrgetter('is_postgresql_9_6'))
     supports_table_partitions = property(operator.attrgetter('is_postgresql_10'))
+    supports_include_indexes = property(operator.attrgetter('is_postgresql_11'))
+    supports_include_gist_indexes = property(operator.attrgetter('is_postgresql_12'))
```
Comment (pope1ni) [inline django/db/backends/postgresql/features.py]: :+1: Well spotted that GIST gained support in PostgreSQL 12! :eagle:

File: django/db/backends/postgresql/schema.py [hunk 3]
```diff
@@ -192,10 +192,11 @@ def _delete_index_sql(self, model, name, sql=None, concurrently=False):
     def _create_index_sql(
         self, model, fields, *, name=None, suffix='', using='',
         db_tablespace=None, col_suffixes=(), sql=None, opclasses=(),
-        condition=None, concurrently=False,
+        condition=None, concurrently=False, include=None
```
Comment (pope1ni) [inline django/db/backends/postgresql/schema.py]: Please add a trailing comma.

File: django/db/backends/postgresql/schema.py [hunk 4]
```diff
@@ -192,10 +192,11 @@ def _delete_index_sql(self, model, name, sql=None, concurrently=False):
...
     ):
         sql = self.sql_create_index if not concurrently else self.sql_create_index_concurrently
         return super()._create_index_sql(
             model, fields, name=name, suffix=suffix, using=using, db_tablespace=db_tablespace,
             col_suffixes=col_suffixes, sql=sql, opclasses=opclasses, condition=condition,
+            include=include
```
Comment (pope1ni) [inline django/db/backends/postgresql/schema.py]: Please add a trailing comma.

File: django/db/models/constraints.py [hunk 5]
```diff
@@ -65,13 +65,16 @@ def deconstruct(self):
 
 
 class UniqueConstraint(BaseConstraint):
-    def __init__(self, *, fields, name, condition=None):
+    def __init__(self, *, fields, name, condition=None, include=()):
```
Comment (pope1ni) [inline django/db/models/constraints.py]: I'd make this `include=None`...

File: django/db/models/constraints.py [hunk 6]
```diff
@@ -65,13 +65,16 @@ def deconstruct(self):
...
         if not fields:
             raise ValueError('At least one field is required to define a unique constraint.')
         if not isinstance(condition, (type(None), Q)):
             raise ValueError('UniqueConstraint.condition must be a Q instance.')
+        if include and not isinstance(include, (list, tuple)):
```
Comment (pope1ni) [inline django/db/models/constraints.py]: ...and then this can be:
```python
if isinstance(include, (type(None), list, tuple)):
```
That is consistent with how `condition` is implemented.

File: django/db/models/constraints.py [hunk 7]
```diff
@@ -65,13 +65,16 @@ def deconstruct(self):
...
+            raise ValueError('UniqueConstraint.include must be a list or tuple.')
         self.fields = tuple(fields)
         self.condition = condition
+        self.include = tuple(include) if include else None
```
Comment (pope1ni) [inline django/db/models/constraints.py]: Make this `tuple(include or ())` and then you don't need `[... in self.include or []]` in the list comprehensions below.

File: django/db/models/indexes.py [hunk 8]
```diff
@@ -11,7 +11,16 @@ class Index:
...
-    def __init__(self, *, fields=(), name=None, db_tablespace=None, opclasses=(), condition=None):
+    def __init__(
+        self,
+        *,
+        fields=(),
+        name=None,
+        db_tablespace=None,
+        opclasses=(),
+        condition=None,
+        include=(),
```
Comment (pope1ni) [inline django/db/models/indexes.py]: Maybe this should be synchronized with the previous cases and also be `None`, along with the related changes in the rest of the `__init__()` method below.

File: docs/ref/models/constraints.txt [hunk 9]
```diff
@@ -115,3 +115,20 @@ ensures that each user only has one draft.
 
 These conditions have the same database restrictions as
 :attr:`Index.condition`.
+
+``include``
+-------------
```
Comment (pope1ni) [inline docs/ref/models/constraints.txt]: Please fix the heading underlines to be the same length as the heading itself.

File: docs/ref/models/constraints.txt [hunk 10]
```diff
@@ -115,3 +115,20 @@ ensures that each user only has one draft.
...
+
+.. attribute:: UniqueConstraint.include
+
+.. versionadded:: 3.1
+
+A list of fields to be included with the unique index. This allows for fully
```
Comment (pope1ni) [inline docs/ref/models/constraints.txt]: `...included with...` → `...include as non-key columns on...`

File: docs/ref/models/constraints.txt [hunk 11]
```diff
@@ -115,3 +115,20 @@ ensures that each user only has one draft.
...
+covered indexes to be used by filtering on the unique field and selecting
+the included fields with :meth:`~django.db.models.query.QuerySet.only()`.
```
Comment (pope1ni) [inline docs/ref/models/constraints.txt]: Perhaps we should have an example of using `QuerySet.only()`?

File: docs/ref/models/indexes.txt [hunk 12]
```diff
@@ -141,3 +141,35 @@ indexes records with more than 400 pages.
 
     The ``condition`` argument is ignored with MySQL and MariaDB as neither
     supports conditional indexes.
+
+``include``
+-------------
```
Comment (pope1ni) [inline docs/ref/models/indexes.txt]: Please fix the heading underlines to be the same length as the heading itself.

File: docs/ref/models/indexes.txt [hunk 13]
```diff
@@ -141,3 +141,35 @@ indexes records with more than 400 pages.
...
+
+.. attribute:: Index.include
+
+.. versionadded:: 3.1
+
+Used to create covering indexes where all columns in a query is available in
+the index. This reduces the number I/O operations needed to run a query.
+
+For example, ``Index(name='covering_index', fields=['last_name'],
+include=['first_name'])`` will allow filtering on ```last_name`` while also
```
Comment (pope1ni) [inline docs/ref/models/indexes.txt]: Fix triple backtick to double backtick...

File: docs/ref/models/indexes.txt [hunk 14]
```diff
@@ -141,3 +141,35 @@ indexes records with more than 400 pages.
...
+query can be produced.
+
+Using ``include`` will produce a smaller index than using a regular
+multi-column index but with the drawback that the columns can not be used as
+predicates or for sorting and still have the benefits of regular indexed
+columns.
+
+:attr:`Index.name` is required when using ``include``.
+
+``include`` are ignored for databases besides PostgreSQL.
```
Comment (pope1ni) [inline docs/ref/models/indexes.txt]: `are` → `is`.

File: docs/releases/3.1.txt [hunk 15]
```diff
@@ -233,6 +236,7 @@ backends.
 * ``DatabaseOperations.fetch_returned_insert_columns()`` now requires an
   additional ``returning_params`` argument.
 
+
```
Comment (pope1ni) [inline docs/releases/3.1.txt]: Drop this unrelated whitespace change.

File: tests/constraints/tests.py [hunk 16]
```diff
@@ -142,6 +142,30 @@ def test_eq_with_condition(self):
             ),
         )
 
+    def test_eq_with_include(self):
+        self.assertEqual(
+            models.UniqueConstraint(
+                fields=['foo', 'bar'], name='include',
+                include=['baz']
```
Comment (pope1ni) [inline tests/constraints/tests.py]: Please add trailing commas or better yet, these can all be condensed to one line I think.

File: django/db/backends/base/schema.py [hunk 17]
```diff
@@ -1087,14 +1100,20 @@ def create_unique_name(*args, **kwargs):
...
+                Statement(
+                    self.sql_create_unique_index,
+                    table=table,
+                    name=name,
+                    columns=columns,
+                    condition=" WHERE " + condition if condition else "",
+                    include=self._get_include_sql(model, include),
+                )
+                if self.connection.features.supports_partial_indexes or
+                self.connection.features.supports_include_indexes
```
[the remark this answers is not in the document]
Comment (hannseman) [reply django/db/backends/base/schema.py]: Good catch 👍 

I've added `_get_index_condition_sql` in a separate commit. I guess we want it separate later when squashing the rest?

File: docs/ref/models/indexes.txt [hunk 18]
```diff
@@ -141,3 +141,35 @@ indexes records with more than 400 pages.
...
+
+``include``
+-------------
+
+.. attribute:: Index.include
+
+.. versionadded:: 3.1
+
+Used to create covering indexes where all columns in a query is available in
+the index. This reduces the number I/O operations needed to run a query.
```
[the remark this answers is not in the document]
Comment (hannseman) [reply docs/ref/models/indexes.txt]: I've tried to make it a bit clearer.

File: docs/releases/3.1.txt [hunk 19]
```diff
@@ -176,6 +176,9 @@ Models
   * ``TREE`` format on MySQL 8.0.16+,
   * ``analyze`` option on MySQL 8.0.18+ and MariaDB.
 
+* Added support for creating covering indexes by using
+  (:attr:`.Index.include`).
```
[the remark this answers is not in the document]
Comment (hannseman) [reply docs/releases/3.1.txt]: I've added a note on `BTreeIndex` and `GistIndex`.

**Comment by pope1ni** [comment]
@hannseman `BloomIndex` was added in #12004 which has now been merged. Could you also add the `self.include` check to `BloomIndex.check_supported()` as covering indexes are not supported there either. Thanks.

File: docs/releases/3.1.txt [hunk 20]
```diff
@@ -203,6 +203,8 @@ Models
   :attr:`~django.db.models.ForeignKey.on_delete` argument of ``ForeignKey`` and
   ``OneToOneField`` emulates the behavior of the SQL constraint ``ON DELETE
   RESTRICT``.
+* Added support for creating covering indexes by using
```
Comment (atombrella) [inline docs/releases/3.1.txt]: Please add a blank line between entries.

File: tests/schema/tests.py [hunk 21]
```diff
@@ -3114,3 +3115,94 @@ class Meta:
...
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+
+        # Define the constraint
+        name = 'unique_partial_constraint_inc'
+        constraint = UniqueConstraint(
+            fields=['uuid'],
+            name=name,
+            include=['height', 'weight'],
+            condition=Q(height__gte=200)
```
[the remark this answers is not in the document]
Comment (hannseman) [reply tests/schema/tests.py]: I think I got all of them now!

File: django/db/backends/base/features.py [hunk 22]
```diff
@@ -294,6 +294,9 @@ class BaseDatabaseFeatures:
     # Does the backend support boolean expressions in the SELECT clause?
     supports_boolean_expr_in_select_clause = True
 
+    # Does the backend support indexes with non-key columns using the INCLUDE clause?
+    supports_include_indexes = False
```
[the remark this answers is not in the document]
Comment (hannseman) [reply django/db/backends/base/features.py]: I could probably need some help with the wording here but I've added something to start working from.

File: tests/postgres_tests/test_indexes.py [hunk 23]
```diff
@@ -442,3 +520,11 @@ def test_spgist_parameters(self):
         with connection.schema_editor() as editor:
             editor.remove_index(CharFieldModel, index)
         self.assertNotIn(index_name, self.get_constraints(CharFieldModel._meta.db_table))
+
+    def test_spgist_include_not_supported(self):
+        index_name = 'spgist_include_exception'
+        index = SpGistIndex(fields=['field'], name=index_name, include=['field2'])
+        with self.assertRaisesMessage(NotSupportedError, 'Covering indexes are not supported for this index type.'):
```
[the remark this answers is not in the document]
Comment (hannseman) [reply tests/postgres_tests/test_indexes.py]: I can agree that the message should be broken out into a variable if it doesn't fit within the line length but I couldn't find any asserts where that is an issue.

File: django/db/backends/base/schema.py [hunk 24]
```diff
@@ -1077,7 +1088,19 @@ def _unique_sql(self, model, fields, name, condition=None):
             'constraint': constraint,
         }
 
-    def _create_unique_sql(self, model, columns, name=None, condition=None):
+    def _get_index_condition_sql(self, condition):
+        if not condition or not self.connection.features.supports_partial_indexes:
+            return ''
+        return ' WHERE ' + condition
+
+    @property
```
[the remark this answers is not in the document]
Comment (hannseman) [reply django/db/backends/base/schema.py]: Sure, can't think of any reason why not.

File: tests/indexes/tests.py [hunk 25]
```diff
@@ -414,3 +463,139 @@ def test_is_null_condition(self):
...
+        with connection.schema_editor() as editor:
+            editor.add_index(Article, index)
+            sql = index.create_sql(Article, editor)
+
+        # Ensure the index is there
+        constraints = connection.introspection.get_constraints(
+            cursor=connection.cursor(), table_name=Article._meta.db_table,
+        )
+        self.assertIn(index_name, constraints)
+        self.assertTrue(
```
[the remark this answers is not in the document]
Comment (hannseman) [reply tests/indexes/tests.py]: You're right, much nicer.

**Comment by atombrella** [comment]
@hannseman Could you please go through the last `assertLess` and `assertIn` cases, to collapse them to possibly use `assertEqual` instead? And then squash the commits. The ticket is already in the [review queue](https://code.djangoproject.com/query?status=!closed&needs_better_patch=0&needs_tests=0&needs_docs=0&has_patch=1&stage=Accepted&desc=1&order=changetime).

**Comment by hannseman** [comment]
@atombrella many thanks, totally missed those `assertLess` and `assertIn ` that were left. 

Squashed the commits and kept https://github.com/django/django/pull/11991/commits/132278b424b03f18bf81b1d2fa50d1c556df9c40 as a separate commit as requested in https://github.com/django/django/pull/11991#discussion_r340668028.

File: django/db/backends/base/schema.py [hunk 26]
```diff
@@ -1077,6 +1077,11 @@ def _unique_sql(self, model, fields, name, condition=None):
             'constraint': constraint,
         }
 
+    def _get_index_condition_sql(self, condition):
+        if not condition or not self.connection.features.supports_partial_indexes:
+            return ''
+        return ' WHERE ' + condition
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: `_get_index_condition_sql()` in the current form don't work properly because it hides database exceptions when we try to create partial indexes on databases which don't support them, it will create a normal index without a condition. This is not an expected behavior. We could create `_index_condition_sql()` without checking `self.connection.features.supports_partial_indexes`, e.g.:
```python
    def _index_condition_sql(self, condition):
        if condition:
            return ' WHERE ' + condition
        return ''
```
but TBH I don't see much value in it.

File: django/db/backends/base/schema.py [hunk 27]
```diff
@@ -942,9 +943,17 @@ def _get_index_tablespace_sql(self, model, fields, db_tablespace=None):
             return ' ' + self.connection.ops.tablespace_sql(db_tablespace)
         return ''
 
+    def _get_index_include_sql(self, model, columns):
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: `_get_index_include_sql` -> `_index_include_sql()`.

File: django/db/backends/base/schema.py (hunk 26 above)
Comment (felixxm) [reply django/db/backends/base/schema.py]: I have the same comment to the `_get_index_include_sql()`, this shouldn't check `supports_include_indexes` feature, IMO.

Comment (felixxm) [reply django/db/backends/base/schema.py]: OK, it makes sense to add a simple `_index_condition_sql()` if we take into account this PR:
```python
def _index_condition_sql(self, condition):
    if condition:
        return ' WHERE ' + condition
    return ''
```

Comment (hannseman) [reply django/db/backends/base/schema.py]: I agree that it's better to crash with a database error if one tries to add an index where these features isn't supported. 

I'll remove these checks.

Comment (hannseman) [reply django/db/backends/base/schema.py]: I've removed the `self.connection.features.supports_partial_indexes` and `self.connection.features.supports_include_indexes`. Some testcases needed to be updated as they where built expecting the silent fallback on non-supported features.

Comment (hannseman) [reply django/db/backends/base/schema.py]: And idea might be to add a check for feature support on `Index` and `UniqueConstraint` similar to how `PostgresIndex.check_supported` is implemented.

Comment (felixxm) [reply django/db/backends/base/schema.py:1094]: > I've removed the `self.connection.features.supports_partial_indexes` and `self.connection.features.supports_include_indexes` checks. Some testcases needed to be updated as they where built expecting the silent fallback on non-supported features.

We added system checks for partial indexes support (see 53d229ff632c2a3e547f2820a94239f38ba4d4ac), so we should restore the previous behavior.

IMO we can ignore indexes with `INCLUDE` if they're not support and add a similar checks for them (check can be added in a separate PR/commit).

I moved `BaseDatabaseSchemaEditor._index_condition_sql()` to #12787.

Comment (hannseman) [reply django/db/backends/base/schema.py]: I've rebased the branch and removed the commit that was added in #12787. The test failure seems like an unrelated flaky test.

**Comment by felixxm** [comment]
@hannseman Can you add a system check for non-local/invalid field names in `include`? (calling `_check_local_fields()` on `include` should be enough, see 3c7bf39e23fe9168f31132d929c9877c5835859b).

**Comment by hannseman** [comment]
@felixxm Thanks for your fixups!

**Comment by felixxm** [comment]
I pushed edits to docs.

**Comment by felixxm** [comment]
I added system checks for covering indexes and unique constraints support.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: docs/ref/models/constraints.txt [hunk 28]
```diff
@@ -145,3 +145,24 @@ enforced immediately after every command.
...
     <https://www.postgresql.org/docs/current/sql-createtable.html#id-1.9.3.85.9.4>`_.
+
+``include``
+-----------
+
+.. attribute:: UniqueConstraint.include
+
+.. versionadded:: 3.2
+
+A list or tuple of the name of the fields to be included in the covering unique
```
Comment (carltongibson) [inline docs/ref/models/constraints.txt:156]:
```suggestion
A list or tuple of the names of the fields to be included in the covering unique
```

File: docs/ref/models/constraints.txt [hunk 29]
```diff
@@ -145,3 +145,24 @@ enforced immediately after every command.
...
+index as non-key columns. This allows index-only scans to be used for queries
+that select only included fields (:attr:`~UniqueConstraint.include`) and filter
+only by unique fields (:attr:`~UniqueConstraint.fields`).
+
+For example::
+
+    UniqueConstraint(name='unique_booking', fields=['room', 'date'], include=['full_name'])
+
+will allow filtering on ``room`` and ``date``, and also selecting ``full_name``
+with having to only fetch data from the index.
```
Comment (carltongibson) [inline docs/ref/models/constraints.txt:166]:
```suggestion
will allow filtering on ``room`` and ``date``, also selecting ``full_name``,
while fetching data only from the index.
```

File: docs/ref/models/indexes.txt [hunk 30]
```diff
@@ -137,3 +137,40 @@ indexes records with more than 400 pages.
...
     supports conditional indexes.
+
+``include``
+-----------
+
+.. attribute:: Index.include
+
+.. versionadded:: 3.2
+
+A list or tuple of the name of the fields to be included in the covering index
```
Comment (carltongibson) [inline docs/ref/models/indexes.txt:148]:
```suggestion
A list or tuple of the names of the fields to be included in the covering index
```

File: docs/ref/models/indexes.txt [hunk 31]
```diff
@@ -137,3 +137,40 @@ indexes records with more than 400 pages.
...
+as non-key columns. This allows index-only scans to be used for queries that
+select only included fields (:attr:`~Index.include`) and filter only by indexed
+fields (:attr:`~Index.fields`).
+
+For example::
+
+    Index(name='covering_index', fields=['headline'], include=['pub_date'])
+
+will allow filtering on ``headline`` and also selecting ``pub_date`` with
+having to only fetch data from the index.
```
Comment (carltongibson) [inline docs/ref/models/indexes.txt:158]:
```suggestion
will allow filtering on ``headline``, also selecting ``pub_date``, while
fetching data only from the index.
```

File: docs/ref/models/constraints.txt [hunk 32]
```diff
@@ -145,3 +145,24 @@ enforced immediately after every command.
 
     Deferred unique constraints may lead to a `performance penalty
     <https://www.postgresql.org/docs/current/sql-createtable.html#id-1.9.3.85.9.4>`_.
+
+``include``
```
Comment (carltongibson) [inline docs/ref/models/constraints.txt:149]: Is include supported just on Postgres? Is that said anywhere, other than 3.2.txt?

Comment (felixxm) [reply docs/ref/models/constraints.txt:149]: I added note:
```
``include`` is supported only on PostgreSQL.
```

**Comment by hannseman** [comment]
@felixxm @charettes @atombrella @pope1ni @carltongibson many thanks for your reviews!
